rfcs/005 - Unsurprising Functions and Scalability RFC.md

```diff
@@ -0,0 +1,215 @@
+
+**Pros:**
+
+- Still pretty fast and we can reuse a lot of existing code.
+
+**Cons:**
+
+- The RPC API has to be reworked again to tackle this.
+- We have to "capture" TypeORM's actions to serialize them to the RPC function, which may be complicated (though our context object should help).
+- We may have to split more complex RPC calls, like `iasql_install` into a collection of multiple smaller ones, (as querying current module state, generating schema, and then populating said schema progressively alter the database state and capturing SQL output from TypeORM to not write into the database immediately may cause future TypeORM SQL statements to differ from what it *should* do due to the state staying frozen).
+
+##### RPC with SQL injection only
+
+Instead of moving the RPC logic into an extension, we could just modify the RPC contract to provide SQL to inject through the current graphile worker approach.
+
+**Pros:**
+
+- No need to introduce Rust into the project
+
+**Cons:**
+
+- Probably even slower than the current approach because the SQL statements returned need to be re-parsed and re-executed after they are received
+- Consumes *far* more data in the database as the SQL statements are kept for potentially months until eventually dropped from the job table.
+- More likely to need a split of complex RPC calls into multiple sub-calls, as no good way to establish a true socket-like communication channel.
+
+##### Extension attaching a second PSQL socket to process
+
+Instead of recreating the RPC mechanism or emulating it, the IaSQL functions call an extension that does two things:
+
+1. Spin up a random port exposing a second PSQL protocol socket.
+2. Calling a REST endpoint on the engine passing the function called and arguments, along with the new socket number.
+
+The extension then polls for:
+
+1. A connection being established on that socket.
+2. Executing whatever SQL comes in on that socket (perhaps as simple as passing this off to Postgres' own internals).
+3. The connection closing.
+
+Then it disables the socket and simply exits the extension function (the wrapper function that called it does what it wants after that to potentially return metadata results to the caller).
+
+**Pros:**
+
+- The `scheduler` process is no longer necessary, but we still get to re-use the vast majority of our current code.
+- No weird injection in TypeORM to capture it's output, just configuring it to talk to a different port than normal.
+
+**Cons:**
+
+- Need to write a Rust extension, and faking the standard postgres connection may be more difficult than expected (it may need to interpret and inject into the main connection)
+- Need to expose all ports between the Postgres server and the IaSQL engine, and RDS cannot be used
+
+#### Change Postgres Behavior
+
+##### Fork Postgres, make it like MySQL
```

Comment:
we could also port everything to mysql instead of postgres. in theory, typeorm is an abstraction on top of both

Reply:
We would need to dig in a bit on that, but I'm not sure we'd have something like `dblink` to let us break stored procedure behavior?

There are also several other downsides to MySQL (for us, at least) that I think make it not worth pursuing, but let me know if you disagree on any of these:

1. MySQL's DB type system is looser, so several data-correctness enforcements we have in the database would need to be moved into the engine itself, which increases development load for us and reduces immediate feedback on bad `INSERT`s for users. For instance in MySQL [ENUMs accept garbage input and batch it all into a singular bucket instead of failing outright](https://dev.mysql.com/doc/refman/5.7/en/enum.html)
2. MySQL doesn't have an array type, so all uses of arrays currently would need to be converted to JSON (which is not typed at all). This is similar to the first point, but rather than being about weird auto-conversion of bad data into a fixed type and cause unexpected behavior, the bad data stays in its original form and could crash our code, instead.
3. Postgres has some types that help us better represent the data we're working with, like the CIDR type, and also allows us to create types to do things better, with the [composite types](https://www.postgresql.org/docs/9.6/rowtypes.html) that we can use where it makes sense for the data to be represented in a clearer way to end users.
4. MySQL is controlled by one of the more capricious corporations out there: Oracle, the company that sued to make API *definitions* copyrightable, which would have effectively destroyed software development in the US. So there is some amount of risk building in that ecosystem if we are very successful, as we don't have the legal budget of Google to fend them off.